common4j/src/main/com/microsoft/identity/common/java/providers/oauth2/AuthorizationResultFactory.java

```diff
@@ -0,0 +1,148 @@
+import com.microsoft.identity.common.java.util.StringUtil;
+import com.microsoft.identity.common.java.util.UrlUtil;
+
+import java.util.Map;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.NonNull;
+
+import static com.microsoft.identity.common.java.AuthenticationConstants.AAD.UPN_TO_WPJ_KEY;
+
+/**
+ * Abstract Factory class which can be extended to construct provider specific {@link AuthorizationResult}.
+ */
+
+// Suppressing rawtype warnings due to the generic types AuthorizationResult and AuthorizationRequest
+@SuppressWarnings(WarningType.rawtype_warning)
+public abstract class AuthorizationResultFactory
+        <GenericAuthorizationResult extends AuthorizationResult,
+                GenericAuthorizationRequest extends AuthorizationRequest> {
+
+    private static final String TAG = AuthorizationResultFactory.class.getSimpleName();
+
+    /**
+     * Factory method which can implemented to construct provider specific {@link AuthorizationResult}.
+     *
+     * @return {@link AuthorizationResult}
+     */
+    public GenericAuthorizationResult createAuthorizationResult(@NonNull final RawAuthorizationResult data,
+                                                                @NonNull final GenericAuthorizationRequest request) {
+
+        final String url = data.getAuthorizationFinalUrl();
+
+        switch (data.getResultCode()) {
+            case CANCELLED:
+                Logger.info(TAG, null, "The authorization request was intentionally cancelled.");
+                return createAuthorizationResultWithErrorResponse(
+                        AuthorizationStatus.USER_CANCEL,
+                        MicrosoftAuthorizationErrorResponse.USER_CANCEL,
+                        MicrosoftAuthorizationErrorResponse.USER_CANCELLED_FLOW
+                );
+
+            case SDK_CANCELLED:
+                Logger.info(TAG, null, "SDK cancelled the authorization request.");
+                return createAuthorizationResultWithErrorResponse(
+                        AuthorizationStatus.SDK_CANCEL,
+                        MicrosoftAuthorizationErrorResponse.SDK_AUTH_CANCEL,
+                        MicrosoftAuthorizationErrorResponse.SDK_CANCELLED_FLOW
+                );
+
+            case COMPLETED:
+                if (StringUtil.isNullOrEmpty(url)) {
+                    Logger.warn(TAG, null, "returned URL is null or empty.");
+                    return createAuthorizationResultWithErrorResponse(
+                            AuthorizationStatus.FAIL,
+                            MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                            MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+                }
+                return parseUrlAndCreateAuthorizationResult(url, request.getState());
+
+            case NON_OAUTH_ERROR:
+                // This is purely client side error, possible return could be chrome_not_installed or the request intent is
```

Comment:
chrome_not_installed I believe is for browser case? What about webview scenario?

Reply:
as of now, the only place that seems to throw this is the PKeyAuthChallengeFactory.